include/seqan3/range/decorator/gap_decorator.hpp

```diff
@@ -1038,6 +655,390 @@ template <std::ranges::view urng_t>
 gap_decorator(urng_t range) -> gap_decorator<urng_t>;
 //!\}
 
+/*!\brief A special iterator that enforces std::random_access_iterator albeit its complexity is logarithmic.
+ *
+ * \details
+ *
+ * \attention This iterator models std::random_access_iterator but **does not guarantee constant time access**.
+ *            Random access is provided in \f$O(\log k)\f$ time (where `k` is the number of gaps).
+ *
+ * This iterator returns values when dereferenced, not references, i.e. it does not model the
+ * [Cpp17random_access_iterator](https://en.cppreference.com/w/cpp/named_req/random_access_iterator)
+ * requirements of the STL. It does model the C++20 std::random_access_iterator.
+ *
+ * You can get a view over the gap_decorator with this random access iterator by using
+ * seqan3::views::pseudo_random_access. But note that after adapting the gap_decorator the non-range interfaces
+ * are lost (e.g. seqan3::gap_decorator::insert_gap or seqan3::gap_decorator::erase_gap).
+ * This adaptor allows using the gap_decorator as a random access range in algorithms like std::ranges::distance and get
+ * the more efficient logarithmic implementation. The regular iterator only models std::bidirectional_iterator
+ * and thus linear time would be needed to calculate the distance.
+ *
+ * \see seqan3::gap_decorator::gap_decorator_iterator
+ */
+template <std::ranges::viewable_range inner_type>
+//!\cond
+    requires std::ranges::random_access_range<inner_type> && std::ranges::sized_range<inner_type> &&
```

Comment:
do we really need all the requires here again? They are probably already checked when instantiating the class

Reply:
Yes it is required so that Compiler knows that it is the same class.

Reply:
Maybe then it's better to move these requirements into the class using static asserts. They need to be checked only once and I think we wanted to do this anyway if requirements do not participate in overload resolution.

Reply:
not resolved

Reply:
No, we can't do it. See I added the handy unary type trait is_instantiable_with to check if I can instantiate the class template with a  specific type. But for this to work I need to constraint the template parameter. static asserts in the body won't work this way.

Reply:
uff... this will be a problem for every type now then that you'd like to test with `is_instantiable_with` because so far we modeled all the requirements via static asserts :/

Reply:
yes. Because we had a wrong assumption on what we can do with it. Having the static_assertion in addition shouldn't matter, though. However, this doesn't need to be that strict. This is a detail function used in a few places only. But handy for such cases.

Reply:
Yes. That's true but unfortunately kind of limiting for fun like this.

Reply:
@smehringer I mark this as resolved. Please unresolve if you think otherwise.